src/Infrastructure/LeanCode.CollectionExtensions/CollectionExtensions.cs

```diff
@@ -0,0 +1,122 @@
+        this IEnumerable<T> enumerable,
+        Func<T, bool> predicate,
+        bool condition
+    )
+    {
+        return condition ? enumerable.Where(predicate) : enumerable;
+    }
+
+    public static IOrderedQueryable<TSource> OrderBy<TSource, TKey>(
+        this IQueryable<TSource> queryable,
+        Expression<Func<TSource, TKey>> selector,
+        bool descending
+    )
+    {
+        return descending ? queryable.OrderByDescending(selector) : queryable.OrderBy(selector);
+    }
+
+    public static IOrderedQueryable<TSource> ThenBy<TSource, TKey>(
+        this IOrderedQueryable<TSource> queryable,
+        Expression<Func<TSource, TKey>> selector,
+        bool descending
+    )
+    {
+        return descending ? queryable.ThenByDescending(selector) : queryable.ThenBy(selector);
+    }
+
+    public static IOrderedEnumerable<TSource> OrderBy<TSource, TKey>(
+        this IEnumerable<TSource> enumerable,
+        Func<TSource, TKey> selector,
+        bool descending
+    )
+    {
+        return descending ? enumerable.OrderByDescending(selector) : enumerable.OrderBy(selector);
+    }
+
+    public static IOrderedEnumerable<TSource> OrderBy<TSource, TKey>(
+        this IEnumerable<TSource> enumerable,
+        Func<TSource, TKey> selector,
+        bool descending,
+        IComparer<TKey> comparer
+    )
+    {
+        return descending ? enumerable.OrderByDescending(selector, comparer) : enumerable.OrderBy(selector, comparer);
+    }
+
+    public static IOrderedEnumerable<TSource> ThenBy<TSource, TKey>(
+        this IOrderedEnumerable<TSource> enumerable,
+        Func<TSource, TKey> selector,
+        bool descending
+    )
+    {
+        return descending ? enumerable.ThenByDescending(selector) : enumerable.ThenBy(selector);
+    }
+
+    public static IQueryable<TResult> LeftJoin<TLeft, TRight, TKey, TResult>(
+        this IQueryable<TLeft> left,
+        IQueryable<TRight> right,
+        Expression<Func<TLeft, TKey>> leftKeySelector,
+        Expression<Func<TRight, TKey>> rightKeySelector,
+        Expression<Func<TLeft, TRight?, TResult>> resultSelector
+    )
+    {
+        var g = Expression.Parameter(typeof(Grouping<TLeft, TRight>), "g");
+        var r = Expression.Parameter(typeof(TRight), "r");
+
+        var keyAccessor = Expression.Property(g, Grouping<TLeft, TRight>.KeyPropertyInfo);
+        var selectorInvocation = Expression.Invoke(resultSelector, keyAccessor, r);
+
+        var lambda = Expression.Lambda<Func<Grouping<TLeft, TRight>, TRight?, TResult>>(selectorInvocation, g, r);
+
+        return left.GroupJoin(
+                right,
+                leftKeySelector,
+                rightKeySelector,
+                (l, r) => new Grouping<TLeft, TRight> { Key = l, Values = r, }
+            )
+            .SelectMany(g => g.Values.DefaultIfEmpty(), lambda);
+    }
+
+    public static IEnumerable<TResult> LeftJoin<TLeft, TRight, TKey, TResult>(
```

Comment:
I'm not sure we should have this - you might use it by accident and I haven't had to use `LeftJoin` on `IEnumerable` ever (although I might be biased).

Reply:
:slightly_frowning_face:

Reply:
Could you express what you have in mind @Saancreed ?

Reply:
Okay, there _are_ cases where this might come in handy, so let's keep it.